### Problem 2

In [32]:
#long term interest rates equal the average of expected future short term rates

In [33]:
#assume short term rates follow random walk

In [34]:
import pandas_datareader as pdr
import pandas as pd
import datetime
import numpy as np
import math
import matplotlib.pyplot as plt
import statsmodels.formula.api as smf
from statsmodels.tsa.stattools import acf
from statsmodels.graphics.tsaplots import plot_acf
from statsmodels.tsa.stattools import acovf

In [35]:
start = datetime.datetime (1960, 1, 1)
end = datetime.datetime (2022, 1, 31)

gs10 = pdr.DataReader(['GS10'], 'fred', start, end)
gs1 = pdr.DataReader(['GS1'], 'fred', start, end)
fedfunds = pdr.DataReader(['FEDFUNDS'], 'fred', start, end)

In [36]:
fedfunds['lag'] = fedfunds.shift(1)

In [38]:
fedfunds

FEDFUNDS   lag
DATE                      
1960-01-01      3.99   NaN
1960-02-01      3.97  3.99
1960-03-01      3.84  3.97
1960-04-01      3.92  3.84
1960-05-01      3.85  3.92
...              ...   ...
2021-09-01      0.08  0.09
2021-10-01      0.08  0.08
2021-11-01      0.08  0.08
2021-12-01      0.08  0.08
2022-01-01      0.08  0.08

[745 rows x 2 columns]

In [39]:
#combine gs10

new_df = pd.merge(left = fedfunds, right = gs10, left_index = True, right_index = True, how = "inner")
new_df.head()

FEDFUNDS   lag  GS10
DATE                            
1960-01-01      3.99   NaN  4.72
1960-02-01      3.97  3.99  4.49
1960-03-01      3.84  3.97  4.25
1960-04-01      3.92  3.84  4.28
1960-05-01      3.85  3.92  4.35

In [40]:
#combine gs1

new_df2 = pd.merge(left = new_df, right = gs1, left_index = True, right_index = True, how = "inner")
new_df2.head()


FEDFUNDS   lag  GS10   GS1
DATE                                  
1960-01-01      3.99   NaN  4.72  5.03
1960-02-01      3.97  3.99  4.49  4.66
1960-03-01      3.84  3.97  4.25  4.02
1960-04-01      3.92  3.84  4.28  4.04
1960-05-01      3.85  3.92  4.35  4.21

### (a)

In [44]:
fedfunds['FEDFUNDS'].mean()

4.834067114093962

In [42]:
gs10.mean()

GS10    5.886148
dtype: float64

In [43]:
gs1.mean()

GS1    4.880134
dtype: float64

### (b)

In [ ]:
#estimate 4 regressions using the data

In [45]:
mod1=smf.ols(formula='GS1~FEDFUNDS',data=new_df2).fit()
mod1.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    GS1   R-squared:                       0.962
Model:                            OLS   Adj. R-squared:                  0.962
Method:                 Least Squares   F-statistic:                 1.895e+04
Date:                Tue, 08 Feb 2022   Prob (F-statistic):               0.00
Time:                        15:38:18   Log-Likelihood:                -750.06
No. Observations:                 745   AIC:                             1504.
Df Residuals:                     743   BIC:                             1513.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.5146      0.040     12.881      0.000       0.436       0.593
FEDFUNDS       0.9031      0.007    137.662      0.000       0.890       0.916
==============================================================================
Omnibus:                      139.494   Durbin-Watson:                   0.299
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              589.910
Skew:                          -0.801   Prob(JB):                    7.99e-129
Kurtosis:                       7.055   Cond. No.                         10.2
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [46]:
mod2=smf.ols(formula='GS10~FEDFUNDS',data=new_df2).fit()
mod2.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   GS10   R-squared:                       0.825
Model:                            OLS   Adj. R-squared:                  0.825
Method:                 Least Squares   F-statistic:                     3498.
Date:                Tue, 08 Feb 2022   Prob (F-statistic):          3.12e-283
Time:                        15:38:20   Log-Likelihood:                -1216.3
No. Observations:                 745   AIC:                             2437.
Df Residuals:                     743   BIC:                             2446.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      2.3792      0.075     31.850      0.000       2.233       2.526
FEDFUNDS       0.7255      0.012     59.140      0.000       0.701       0.750
==============================================================================
Omnibus:                        1.713   Durbin-Watson:                   0.093
Prob(Omnibus):                  0.425   Jarque-Bera (JB):                1.782
Skew:                           0.111   Prob(JB):                        0.410
Kurtosis:                       2.911   Cond. No.                         10.2
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [48]:
mod3=smf.ols(formula='GS1~FEDFUNDS+lag',data=new_df2).fit(cov_type = 'HC1')
mod3.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    GS1   R-squared:                       0.962
Model:                            OLS   Adj. R-squared:                  0.962
Method:                 Least Squares   F-statistic:                     3045.
Date:                Tue, 08 Feb 2022   Prob (F-statistic):               0.00
Time:                        15:38:49   Log-Likelihood:                -748.29
No. Observations:                 744   AIC:                             1503.
Df Residuals:                     741   BIC:                             1516.
Df Model:                           2                                         
Covariance Type:                  HC1                                         
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.5109      0.046     11.206      0.000       0.422       0.600
FEDFUNDS       0.8657      0.087      9.951      0.000       0.695       1.036
lag            0.0379      0.085      0.445      0.656      -0.129       0.205
==============================================================================
Omnibus:                      137.223   Durbin-Watson:                   0.292
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              566.830
Skew:                          -0.795   Prob(JB):                    8.21e-124
Kurtosis:                       6.970   Cond. No.                         24.1
==============================================================================

Notes:
[1] Standard Errors are heteroscedasticity robust (HC1)
"""

In [31]:
mod4=smf.ols(formula='GS10~FEDFUNDS+lag',data=new_df2).fit(cov_type = 'HC1')
mod4.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   GS10   R-squared:                       0.828
Model:                            OLS   Adj. R-squared:                  0.828
Method:                 Least Squares   F-statistic:                     1046.
Date:                Mon, 07 Feb 2022   Prob (F-statistic):          1.46e-216
Time:                        20:36:05   Log-Likelihood:                -1208.0
No. Observations:                 744   AIC:                             2422.
Df Residuals:                     741   BIC:                             2436.
Df Model:                           2                                         
Covariance Type:                  HC1                                         
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      2.3613      0.070     33.653      0.000       2.224       2.499
FEDFUNDS       0.3903      0.130      3.000      0.003       0.135       0.645
lag            0.3386      0.130      2.597      0.009       0.083       0.594
==============================================================================
Omnibus:                        2.265   Durbin-Watson:                   0.084
Prob(Omnibus):                  0.322   Jarque-Bera (JB):                2.327
Skew:                           0.114   Prob(JB):                        0.312
Kurtosis:                       2.848   Cond. No.                         24.1
==============================================================================

Notes:
[1] Standard Errors are heteroscedasticity robust (HC1)
"""

### (c)

In [ ]:
#test theory that b0 = 1 for GS1 and GS10

In [62]:
import scipy.stats                                                                              
                                                                                       
alpha = 0.05                                                                                    
t_stat = scipy.stats.t.ppf(1 - alpha / 2, len(new_df) - 2)    
t_stat

1.9631619315801032

In [57]:
mod1_test = (mod1.params[1] - 1) / 0.007/(len(new_df))**(1/2)
mod1_test

-0.5072469426202761

In [63]:
mod2_test = (mod2.params[1] - 1) / 0.012/(len(new_df))**(1/2)
mod2_test

-0.8381760289335837

In both cases the null hypothesis fails to be rejected, stating that there is sufficient evidence to prove that b0 = 1. For model 1 (GS1), the calculated t statistic is -.51 which is not less than -1.96. For model 1 (GS10), the calculated t statistic is -.84 which is not less than -1.96. b0 = 1 means that the effect of fed funds on GS1 and GS10 is a permanent shock. 

### (d) 

In [ ]:
#model 2 - is coefficient b1 statistically significant? Instead of b2 on HW

b1 is not statistically significant for model 2 when regressing GS1 on the fed funds rate and the lag rate. This is because the p value is .656 which is greater than .05. 

b1 is statistically significant for model 2 when regressing GS10 on the fed funds rate and the lag rate. This is because the p value is .009 which is less than .05. 

### (e)

In [93]:
#bo + b1 = 1
mod3_t_stat = (mod3.params[1] + mod3.params[2] - 1) / (mod3.bse[1]**2/len(new_df) + mod3.bse[2]**2/len(new_df))**(1/2)
mod3_t_stat

-21.630001033306524

In [86]:
mod4.params[2]

0.33864216953658177

In [95]:
mod3_pvalue = scipy.stats.t.pdf(mod3_t_stat, len(new_df)-2)
mod3_pvalue

4.9896181385905406e-80

In [94]:
#bo + b1 = 1
mod4_t_stat = (mod4.params[1] + mod4.params[2] - 1) / (mod4.bse[1]**2/len(new_df) + mod4.bse[2]**2/len(new_df))**(1/2)
mod4_t_stat

-40.165402228751894

In [96]:
mod4_pvalue = scipy.stats.t.pdf(mod4_t_stat, len(new_df)-2)
mod4_pvalue

1.3854794048390529e-187